File: Java/01_Basics/09_ExceptionHandling.java
// ─────────────────────────────────────────
//  EXCEPTION HANDLING
// ─────────────────────────────────────────

public class ExceptionHandling {

    // ── Custom checked exception ──────────────
    static class InsufficientFundsException extends Exception {
        private final double amount;
        private final double balance;

        public InsufficientFundsException(double amount, double balance) {
            super(String.format("Cannot withdraw %.2f. Balance is only %.2f.", amount, balance));
            this.amount = amount;
            this.balance = balance;
        }

        public double getAmount()  { return amount; }
        public double getBalance() { return balance; }
    }

    // ── Custom unchecked exception ─────────────
    static class ValidationException extends RuntimeException {
        public ValidationException(String message) {
            super(message);
        }
    }

    // ── Checked exception — must declare or handle ─
    static double withdraw(double balance, double amount) throws InsufficientFundsException {
        if (amount <= 0)
            throw new ValidationException("Amount must be positive");
        if (amount > balance)
            throw new InsufficientFundsException(amount, balance);
        return balance - amount;
    }

    public static void main(String[] args) {

        // ── Basic try/catch ───────────────────────
        try {
            int result = 10 / 0;
        } catch (ArithmeticException e) {
            System.out.println("Error: " + e.getMessage());  // / by zero
        }

        // ── Catch multiple exceptions ─────────────
        try {
            int[] arr = {1, 2, 3};
            System.out.println(arr[10]);
        } catch (ArrayIndexOutOfBoundsException e) {
            System.out.println("Index out of bounds!");
        } catch (Exception e) {   // catch-all — put last
            System.out.println("Unexpected: " + e.getMessage());
        }

        // ── Multi-catch (Java 7+) ─────────────────
        try {
            String s = null;
            s.length();
        } catch (NullPointerException | IllegalArgumentException e) {
            System.out.println("Caught: " + e.getClass().getSimpleName());
        }

        // ── finally — always runs ─────────────────
        try {
            Integer.parseInt("abc");
        } catch (NumberFormatException e) {
            System.out.println("Parse error: " + e.getMessage());
        } finally {
            System.out.println("Finally always runs — use for cleanup");
        }

        // ── Custom exception ──────────────────────
        try {
            double newBalance = withdraw(100.0, 150.0);
        } catch (InsufficientFundsException e) {
            System.out.println(e.getMessage());
            System.out.printf("Tried: %.2f, Had: %.2f%n", e.getAmount(), e.getBalance());
        }

        // Unchecked — no need to declare or handle
        try {
            withdraw(100.0, -50.0);
        } catch (ValidationException e) {
            System.out.println("Validation: " + e.getMessage());
        } catch (InsufficientFundsException e) {
            System.out.println("Funds: " + e.getMessage());
        }

        // ── try-with-resources (auto close) ───────
        // Any class implementing AutoCloseable is auto-closed
        try (var reader = new java.io.StringReader("hello")) {
            int c = reader.read();
            System.out.println((char) c);   // h
        } catch (java.io.IOException e) {
            System.out.println(e.getMessage());
        }
        // reader.close() called automatically even on exception

        // ── Checked vs Unchecked ──────────────────
        // Checked (extends Exception):
        //   Must be caught or declared with throws
        //   Examples: IOException, SQLException, ClassNotFoundException
        //
        // Unchecked (extends RuntimeException):
        //   Compiler doesn't force handling — programmer's mistake
        //   Examples: NullPointerException, ArrayIndexOutOfBoundsException,
        //             IllegalArgumentException, ArithmeticException

        // ── Common exceptions reference ────────────
        // ArithmeticException       — e.g. division by zero
        // ArrayIndexOutOfBoundsException — bad array index
        // NullPointerException      — null reference used
        // ClassCastException        — bad cast
        // NumberFormatException     — bad string parse
        // IllegalArgumentException  — bad argument
        // IllegalStateException     — bad object state
        // StackOverflowError        — infinite recursion
        // OutOfMemoryError          — heap full
        // IOException               — I/O failure
        // FileNotFoundException     — file missing
        // SQLException              — database error

        // ── Stack trace ───────────────────────────
        try {
            methodA();
        } catch (RuntimeException e) {
            e.printStackTrace();   // prints full call stack
            System.out.println("Exception type: " + e.getClass().getName());
            System.out.println("Message: " + e.getMessage());
        }
    }

    static void methodA() { methodB(); }
    static void methodB() { throw new IllegalStateException("Something went wrong in B"); }
}

File: CPP/01_Basics/09_ExceptionHandling.cpp
// ─────────────────────────────────────────
//  EXCEPTION HANDLING
// ─────────────────────────────────────────

#include <iostream>
#include <stdexcept>   // standard exceptions
#include <string>
using namespace std;

// ── Custom exception ──────────────────────
class InsufficientFundsException : public runtime_error {
    double amount_;
    double balance_;
public:
    InsufficientFundsException(double amount, double balance)
        : runtime_error("Insufficient funds"),
          amount_(amount), balance_(balance) {}

    double amount()  const { return amount_; }
    double balance() const { return balance_; }

    const char* what() const noexcept override {
        static string msg;
        msg = "Cannot withdraw " + to_string(amount_) +
              ". Balance is only " + to_string(balance_);
        return msg.c_str();
    }
};

class ValidationException : public invalid_argument {
public:
    explicit ValidationException(const string& msg)
        : invalid_argument(msg) {}
};

double withdraw(double balance, double amount) {
    if (amount <= 0)
        throw ValidationException("Amount must be positive");
    if (amount > balance)
        throw InsufficientFundsException(amount, balance);
    return balance - amount;
}

int main() {

    // ── Basic try/catch ───────────────────────
    try {
        int result = 10 / 0;   // undefined behavior in C++ (not an exception!)
        cout << result << endl;
    } catch (...) {
        cout << "Caught something" << endl;
    }
    // Note: integer division by zero is UB in C++, not always caught!
    // Floating-point division by zero gives infinity (not an exception)
    cout << 1.0 / 0.0 << endl;   // inf

    // Throw and catch an int (can throw anything)
    try {
        throw 42;
    } catch (int e) {
        cout << "Caught int: " << e << endl;
    }

    // Throw a string
    try {
        throw string("something went wrong");
    } catch (const string& e) {
        cout << "Caught string: " << e << endl;
    }

    // ── Standard exceptions ───────────────────
    try {
        throw runtime_error("Runtime error occurred");
    } catch (const runtime_error& e) {
        cout << "runtime_error: " << e.what() << endl;
    }

    try {
        vector<int> v = {1, 2, 3};
        cout << v.at(10) << endl;   // throws std::out_of_range
    } catch (const out_of_range& e) {
        cout << "out_of_range: " << e.what() << endl;
    }

    // ── Catch hierarchy ───────────────────────
    // Derived exceptions first, base exceptions last
    try {
        throw out_of_range("Index too large");
    } catch (const out_of_range& e) {    // more specific — catch first
        cout << "out_of_range: " << e.what() << endl;
    } catch (const runtime_error& e) {   // more general
        cout << "runtime_error: " << e.what() << endl;
    } catch (const exception& e) {       // catch-all for std exceptions
        cout << "exception: " << e.what() << endl;
    } catch (...) {                       // catch absolutely everything
        cout << "Unknown exception" << endl;
    }

    // ── Custom exception ──────────────────────
    try {
        withdraw(100.0, 150.0);
    } catch (const InsufficientFundsException& e) {
        cout << e.what() << endl;
        cout << "Tried: " << e.amount() << ", Had: " << e.balance() << endl;
    }

    try {
        withdraw(100.0, -50.0);
    } catch (const ValidationException& e) {
        cout << "Validation: " << e.what() << endl;
    } catch (const InsufficientFundsException& e) {
        cout << "Funds: " << e.what() << endl;
    }

    // ── Re-throw ──────────────────────────────
    auto risky = []() {
        try {
            throw runtime_error("Original error");
        } catch (const runtime_error& e) {
            cout << "Logging: " << e.what() << endl;
            throw;   // re-throws the same exception
        }
    };

    try {
        risky();
    } catch (const runtime_error& e) {
        cout << "Re-caught: " << e.what() << endl;
    }

    // ── noexcept ──────────────────────────────
    // Marks a function as not throwing — optimizer hint + contract
    auto safeAdd = [](int a, int b) noexcept -> int {
        return a + b;
    };
    cout << safeAdd(3, 4) << endl;

    // ── Standard exception hierarchy ──────────
    // exception
    // ├── logic_error
    // │   ├── invalid_argument   — bad argument value
    // │   ├── domain_error       — math domain error
    // │   ├── length_error       — exceeds max size
    // │   └── out_of_range       — value out of range
    // └── runtime_error
    //     ├── range_error        — result out of range
    //     ├── overflow_error     — arithmetic overflow
    //     └── underflow_error    — arithmetic underflow
    // bad_alloc    — new failed (out of memory)
    // bad_cast     — dynamic_cast failed
    // bad_typeid   — typeid on null pointer

    // ── RAII — Resource Acquisition Is Initialization ─
    // C++ idiom: acquire resources in constructor, release in destructor
    // Ensures cleanup even if exceptions occur (no need for finally)
    class Resource {
        string name_;
    public:
        Resource(string name) : name_(name) {
            cout << "Acquired: " << name_ << endl;
        }
        ~Resource() {
            cout << "Released: " << name_ << endl;  // called on scope exit
        }
    };

    try {
        Resource r("database connection");
        throw runtime_error("Something failed");
        // r's destructor is called automatically even here!
    } catch (const exception& e) {
        cout << "Caught: " << e.what() << endl;
    }
    // Output: Acquired → Released → Caught

    return 0;
}
